KEP 2079 - Add PortRange KEP

rikatz opened the pull request:
Signed-off-by: Ricardo Pchevuzinske Katz <ricardo.katz@gmail.com>

This PR adds the KEP for the PortRange addition in NetworkPolicy specification

Issue: #2079

/cc @andrewsykim @jayunit100 @abhiraut 
/assign @thockin @caseydavenport

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+
+API changes to NetworkPolicy:
+* Add a new field ``spec.ingress|egress.ports.PortRange
+* On field creation/change a validation should happen to verify if the format is compliant with ``nnnnn,nnnn1-nnnn5,http`` where nnnnn is a number between 0 and 65535 and named ports aren't mixed with numbered ports
```
I don't think we've had much discussion yet on what the API for this should look like, maybe something to discuss on Monday's call?

I'm a bit opposed to putting the range semantic in the field value like `nnnn1-nnnn5`, I'm wondering if instead we might want explicit fields for `to`, `from` and `except`, something like:
```
- ingress:
  - portRange:
      from: 30000
      to: 32000
      except:
          - 31000
          - 31001
```  

Wondering how exceptions are expressed with ``nnnnn,nnnn1-nnnn5,http``, maybe I missed that.

rikatz replied:
Yeah, I remember there was a discussion somewhere about the syntax, but because I couldn't find I've wrote whatever came to my mind expected you folks to jump in and discuss with me the best syntax :)

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+-->
+
+API changes to NetworkPolicy:
+* Add a new field ``spec.ingress|egress.ports.PortRange
```
Wondering if it makes sense to have `portRange` under `ports` or should `ports` and `portRanges` be separate fields under `ingress` and `egress`? What happens if `portRanges` conflicts with `port`?

rikatz replied:
I think it's up to the CNI how it's going to deal with the conflicts? Or should we validate this in the API?

 But I have no opposition to move portRanges to outside ports, whatever is cleaner/better

abhiraut replied:
if we move port under portRange(or maybe some new name, so having port under portRange doesn't sound awkward) and validate at API level that only one of those fields is set, we can eliminate potential conflicts

rikatz replied:
do we want to eliminate the conflict, or consider only one field? 

Or maybe convert a .port entry to a .portRange.port as proposed in the comment below

Also, about the naming thing, maybe .ports.Range can be a better naming?

andrewsykim replied:
I think it would be useful to see all the example yaml for all the options, maybe in a separate google doc for now?

abhiraut commented:
+1 to what @andrewsykim said regarding the use of string in port range. The following is what i had in mind:
```
- ingress:
  - ports:
     protocol: TCP        # existing field
     port: 6379           # existing field.. internally converted to new field
     portRange:
       from: 30000        # start of the range, must be set if portRange.to is set, cannot be set with portRange.port
       to: 32000          # end of the range, must be set if portRange.from is set, cannot be set with portRange.port
       port: 8080         # single entry, cannot be set with portRange.from and portRange.To, this could be intOrSting to include a named port.
       except:
          - 31000         # no allow rule for this port and must be within the above portRange (wondering if except should also handle a range of port to be excepted?). cannot be a named port
          - 31001
```

andrewsykim commented:
Would it ever make sense to have the port exceptions be a range as well? I'm thinking not but wondering if there are use-cases for this. Are port exceptions generally for a small number of ports in a large range?

abhiraut commented:
> Would it ever make sense to have the port exceptions be a range as well? I'm thinking not but wondering if there are use-cases for this. Are port exceptions generally for a small number of ports in a large range?

yup .. that's not the use case i came across either

rikatz commented:
Maybe there's some case where exceptions should be a port-range either (like allowing the access to any port BUT well known node ports?)

andrewsykim commented:
Seems like this needs a bit more discussion, removing assignees for now until then

/unassign @thockin @caseydavenport

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Today the ``ports``field in ingress and egress network policies is an array that needs a declaration of each single port to be contemplated. This KEP proposes to add a new field that allows a declaration of a port range, simplifying the creation of rules with multiple ports.
```
Please wrap lines to make it easier to review

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+
+NetworkPolicy object is a complex object, that allows a developer to specify what's the traffic behavior expected of the application and allow/deny undesired traffic.
+
+There are a number of user issues like [this](https://github.com/kubernetes/kubernetes/issues/67526) and [this](https://github.com/kubernetes/kubernetes/issues/93111) where users expose the need to create a policy that allow a range of ports but some specific port, or also cases that a user wants to create a policy that allows the egress to other cluster to the NodePort range (eg 32000-32768) and ...
```
Preferred KEP style is to use "reference" style links like the "`[experience reports]`" example in the comment just above, and refer to the issues explicitly (eg "`a number of user issues like [kubernetes #67526]`")

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+
+#### Story 1
+
+As the owner of a web application I want to allow the traffic to my Pods in different ports without having to create multiple network policies. As an example, my application is exposed in ports 80, 443 and 8443 and I want to group this in a single rule
```
You can do that now, you just need to specify

    - ports:
        - port: 80
        - port: 443
        - port: 8443

So why isn't that good enough for this user? (If your answer is "well this is just an example, in other cases you couldn't do this", then throw out this example and replace it with one of those better cases.)

(Also, I assume this means your proposal for `portRange` is going to allow comma-separated ports, but you don't say that anywhere.)

rikatz replied:
You are right. I'm dropping this user story then. About the comma-separated, we've reached to api-machinery which said it's not a good idea as we would need to leave the validation of a string also to the API. https://kubernetes.slack.com/archives/CJUQN3E4T/p1603141881023900

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+
+#### Story 3
+
+As the owner of an application, I want to allow the communication with a set of ports between 6000 and 7000 except port 6667.
```
This is not a user story, it's just a restatement of the proposed feature.

I can write out "As a user I want to give all of my money to @danwinship", but if I don't provide any more details/justification then that, then I wouldn't expect anyone else to believe that there are actually real users out there who need that feature. (But there totally are.)

rikatz replied:
OK, the user story has been removed anyway as it doesn't make sense now

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+
+As the owner of an application, I want to allow the communication with a set of ports between 6000 and 7000 except port 6667.
+
+### Notes/Constraints/Caveats (Optional)
```
As mentioned in one of the issues, this feature would be difficult/inefficient to implement in Open vSwitch-based network plugins (eg, openshift-sdn, ovn-kubernetes, Antrea), since OpenFlow does not have any way to specify port ranges. There may also be other network types that have difficulty implementing this as well.

rikatz replied:
@danwinship I couldn't find the comment in the previous issues, but anyway I'm taking a look into OVS stuff to check the constraint (I believe you, just using this as an excuse to study it better).

If you have the comment can you please provide it so I can feed the document? 

Anyway I'm inserting this as a general caveat, it depends if the technology used by the CNI provider supports multiport. It came to my mind now that this might be a problem as well to eBPF based CNIs that would need to change how they populate the maps used by programs as well, so maybe it would be a need of the CNI to loop through all the ports and generate a rule for each of them.

@aojea also pointed into a [part of code](https://github.com/kubernetes/kubernetes/blob/97a13a0/pkg/security/podsecuritypolicy/provider.go#L421) that converts an existing ``HostPortRange`` field to comma-separated but it seems to me it's mostly related to PSP so wondering how this also could be re-used.

Anyway, I'll point that in the doc.

rikatz replied:
Found it, never mind

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,623 @@
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+#### Removing a Deprecated Flag
```
You should get rid of as much template text as you can to make it easier to see the actual KEP text. eg, just delete any sections (like this one about command-line flags) which are irrelevant to your KEP.

rikatz commented:
@andrewsykim As suggested, created the https://docs.google.com/document/d/1dDYxjsWo17xRq4tqONOA3MBNCqIzUn8AMovCQykE6-8/edit?usp=sharing to review the design

@danwinship Thank you for the review! I'll come back to your comments later this week, once we get some better design. About your concern of 'this feature would be difficult/inefficient to implement in Open vSwitch-based network plugins', you're suggesting this is specified in the "Constraints" section, right?

rikatz commented:
Removed the user story of exceptions also, as this may be "implicit" in the range (you can define a range of 32000-33000 and 33002-34000 dropping the 33001 port)

rikatz commented:
/cc @cmluciano

abhiraut commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,286 @@
+## Proposal
+
+
+In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct 
```
could you also add information on what to expect when both the `Port` and `Range` fields are set? do we allow them to be set simultaneously ? perhaps a clarification about it

rikatz replied:
OK, so here I'm leaving as:

"If both ``Port`` and ``Range`` are specified, they are cumulative and no further
validation might occur. It's up to the CNI to summarize both of the fields in a 
single rule."

I don't know if this is our ideal scenario, but I can't neither see why we shouldn't do this way, after all we're specifying some ports to be allowed, are they in singular field (Port) or range field (Range).

Sounds good?

abhiraut replied:
agree with you.. at least we clarify the usage in KEP so it is not misunderstood

abhiraut commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,286 @@
+
+TBD
+
+## Proposal
```
is there a section for Alternate Proposal, where we can outline the proposal with "Exception" and why we considered it as not a viable option .. for the sake of completeness ?

rikatz replied:
We can maybe add this to Non-Goals, like:

* Specifying Exceptions to the range -> An exception can at this time be specified as multiple ranges not contemplating the exception. Also specifying an exception at this time may lead to much more complex implementation, needing to interact through all the ports[].ranges to see if the exception is inside any of them, and may lead to some confusion.

rikatz replied:
@andrewsykim any thoughts on this also?

danwinship replied:
(mark this conversation resolved?)

danwinship commented:
> About your concern of 'this feature would be difficult/inefficient to implement in Open vSwitch-based network plugins', you're suggesting this is specified in the "Constraints" section, right?

and/or Drawbacks

rikatz commented on keps/sig-network/2079-network-policy-port-range/README.md:
````diff
@@ -0,0 +1,286 @@
+```
+// NetworkPolicyPort describes a port to allow traffic on
+type NetworkPolicyPort struct {
+    Range PortRange
````
changed here, good catch :)

abhiraut commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,296 @@
+## Drawbacks
+
+*  The technology used by the CNI provider might not support port range in a 
+trivial way. As an example, OpenFlow does not have a way to specify port range
```
actually we can leverage the tcp_dst field in ovs in the port/mask format to specify range of ports and reduce the number of flows required. This is documented at https://www.openvswitch.org/support/dist-docs-2.5/ovs-ofctl.8.txt  | grep tcp_dst

Excerpt from the doc:
/*
              Bitwise  matches  on transport ports are rarely useful in isola‐
              tion, but a group of them can be used to reduce  the  number  of
              flows  required  to  match  on  a range of transport ports.  For
              example, suppose that the goal is to match TCP source ports 1000
              to  1999,  inclusive.   One way is to insert 1000 flows, each of
              which matches on a single source port.  Another way is  to  look
              at the binary representations of 1000 and 1999, as follows:
*/

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,306 @@
+
+### Test Plan
+
+TBD
```
This is probably needs to be filled in since the we changed the status to `implementable`. I think something like this would be efficient:

Unit tests:
* test API validation logic
* test API strategy to ensure disabled fields 

E2E tests:
*  Add e2e tests exercising only the API operations for port ranges. Data-path validation should be done by CNIs.

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,306 @@
+type NetworkPolicyPortRange struct {
+    From         uint16
+    To           uint16
+    Exceptions:  []uint16
```
Can this field just be called `Except`?

rikatz replied:
sounds good

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
````diff
@@ -0,0 +1,306 @@
+new struct:
+```
+// NetworkPolicyPort describes a port to allow traffic on
+type NetworkPolicyPort struct {
````
Can we add the whole struct here instead of just the range field?

rikatz replied:
Added. BTW the documentation says:

```
// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
	// +optional
	Port *intstr.IntOrString
```

I'm worried here about if we provide ``Range`` but no ``Port`` it will still match all Ports, so I've changed the struct doc to:

```
// The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided but a Range is 
  // provided, this field is ignored. Otherwise this matches all port names and
  // numbers.
```

This validation will probably need to be changed in the CNI? Or I need to check if this is something from the API

thockin replied:
How about:

```
  // The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided, this matches all port names and
  // numbers, whether an endPort is defined or not.
```

Or were you actually proposing something different?

rikatz replied:
IMO it's good. We're exactly proposing in validations (I think we are!) that endPort can only exist if Port is defined.

Also I think this comment is old, when we were still considering using PortRange (and not EndPort), the new comment was:

```
// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
```

But I've changed as you've proposed, it's clearer

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
````diff
@@ -0,0 +1,306 @@
+```
+// NetworkPolicyPort describes a port to allow traffic on
+type NetworkPolicyPort struct {
+    Range NetworkPolicyPortRange
````
Might be good to add the API docs here now.

andrewsykim replied:
Also I _think_ this has to be a pointer to allow for disabled fields.

rikatz replied:
changed, no idea either

rikatz replied:
Answering myself, yes they should be a pointer: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

rikatz replied:
but otoh, we have ``Number int32`` in ``ServiceBackendPort`` and this is probably because int32 have a default nil (which is 0)? Will try

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,306 @@
+
+#### Beta
+- ``PortRanges`` has been supported for at least 1 minor release
+- At least one CNI provider implements the new field
```
Personally I think only requiring one implementation is too loose of a requirement for this field to be on by default. Not sure what the magic number is but should at least be a handful of CNIs?

rikatz replied:
Yes, it was thinking more about having a PR to be reviewed other then a magic number.

I'm tempted to follow the same number proposed by Jay in #2113  and change to:

```
- The name selector has been supported for at least 1 minor release.
- Four commonly used NetworkPolicy (or CNI providers) implement the new field, 
with generally positive feedback on its usage.
- Feature Gate is enabled by Default.```

rikatz commented:
/assign @thockin

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+The range will need to be validated, with the following scenarios:
+* If there's a ``From`` or a ``To`` field defined, the other one must be defined.
+* ``From`` needs to be less than or equal to ``To``
+* All the ports in the ``Exceptions`` array must be inside the defined range.
```
This is probably a minor thing, but will we reject NetworkPolicys that have duplicate ports in `Except`?

rikatz replied:
I don't think this is actually a big problem.

A similar question: what happens if I create two ``NetworkPolicyPort`` specifying the same port? I've tested here and the API accepts so I think we're good with repeated excepts as well

rikatz replied:
Removed except, closing this conversation :)

andrewsykim commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+  // provided, this field is ignored. Otherwise this matches all port names and
+  // numbers.
+  // +optional
+  Port *intstr.IntOrString
```
I'm still on the fence about this, but I feel that `Port` and `Range` under the same field feels awkward. Putting range semantics into it's own struct _might_ be cleaner, even if you have to specify the Protocol field twice. This way consumers don't have to reason about precedent between `port` and `range` under the same field, but can just take the union of `ports` and `portRanges`.  

```
ports:
  - protocol: TCP
    port: 80
  - protocol: TCP
    port: 443
portRanges:
  - protocol: TCP
    range:
      from: 30000
      to: 32000
```

rikatz replied:
tbh I have no strong opinion about naming either, so I can move this to outside ports, making it an array and that's fine

andrewsykim replied:
Let's wait to see what Tim thinks :)

rikatz replied:
Great, thank you for all the feedback Andrew :)

danwinship replied:
I agree `portRanges` is clearer.

rikatz replied:
@andrewsykim @danwinship about @thockin proposal of just adding an EndPort into the existing ``NetworkPolicyPort`` sounds good and this conversation can be resolved?

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+<!-- toc -->
```
You seem to be missing a title now

rikatz replied:
Added, thanks!

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+validation might occur. It's up to the CNI to summarize both of the fields in a 
+single rule.
+
+### User Stories (Optional)
```
It's odd that you don't have any stories inspired by the two issues linked above.

The single existing user story seems like it could potentially be solved in other ways than allowing arbitrary port ranges.

In general, I feel like this feature is not an obvious must-have for NetworkPolicy, and so I feel like the KEP needs to make a strong case for it. And that means making a case that (a) there are use cases that fit within the design of NetworkPolicy (ie, application-focused policy, _not_ administrator-focused policy) which cannot easily be solved with the current NetworkPolicy features, but which could be solved if we added port ranges, _and_ (b) the port range solution does a better job of solving the use cases than any other obvious alternatives

rikatz replied:
You're right :)

I've added new user stories, trying to make this more clarifying (following Winship's 101 User Story course from your Kep :P ) so let's see if with the new User Stories people get convinced this is necessary.

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+    To           uint16
+
+    // Except defines all the exceptions in the port range
+    Except:      []uint16
```
wait, I thought exceptions were going away?

rikatz replied:
@danwinship Last policy api meeting we've decided that the efforts to add exceptions again is too low compared to the benefits of adding it in the struct.

Basically, it's clearer when you define an exception inside a range instead of defining multiple ranges excluding an exception, like:

```
- from: 32000
  to: 32999
- from: 33001 # Excluding the 33000
  to: 34000
```
vs
```
- from: 32000
  to: 34000
  except:
  - 33001
```

So I've reinserted this in the struct, and removed this as a non goal :)

danwinship replied:
Well, ok, but what is the use case for exceptions then, such that they need to be easy to express? What is the developer use case for granting a pod access to 32000-34000 _except_ 33001?

There were strong objections to adding `except` to `ipBlock`, but we kept it because it's incredibly difficult to do the equivalent of

    - from: 32000
      to: 32999
    - from: 33001
      to: 34000

with CIDRs. if we didn't have `except` then in order to write

    - ipBlock:
        cidr: 192.168.0.0/24
        except:
          - 192.168.0.5

you'd have to do:

    - cidr: 192.168.0.128/25
    - cidr: 192.168.0.64/26
    - cidr: 192.168.0.32/27
    - cidr: 192.168.0.16/28
    - cidr: 192.168.0.8/29
    - cidr: 192.168.0.6/31
    - cidr: 192.168.0.4/32
    - cidr: 192.168.0.2/31
    - cidr: 192.168.0.1/32

which is both horrifying and incomprehensible (and probably also wrong; I didn't double-check it).

rikatz replied:
I think this user story is still compliant, even if we've removed the exception, thinking that the user can create multiple ranges excluding the excepts, right?

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+
+### Validations
+The range will need to be validated, with the following scenarios:
+* If there's a ``From`` or a ``To`` field defined, the other one must be defined.
```
They are not optional fields, as currently defined

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+If upgraded no impact should happen as this is a new field.
+
+If downgraded the CNI wont be able to look into the new field, as this does not 
+exists and network policies using this field will stop working.
```
They will stop working _correctly_ and start working incorrectly.

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+    To           uint16
+
+    // Except defines all the exceptions in the port range
+    Except:      []uint16
```
```suggestion
    Except       []uint16
```

rikatz replied:
Solved, thanks!

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+The range will need to be validated, with the following scenarios:
+* If there's a ``From`` or a ``To`` field defined, the other one must be defined.
+* ``From`` needs to be less than or equal to ``To``
+* All the ports in the ``Exceptions`` array must be inside the defined range.
+* ``Exception`` can only be defined if ``From`` and ``To`` are also defined.
```
```suggestion
* All the ports in the ``Except`` array must be inside the defined range.
* ``Except`` can only be defined if ``From`` and ``To`` are also defined.
```

but again, `From` and `To` are currently required to always be defined

danwinship commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,339 @@
+
+  - API type(s): NetworkPolicy / NetworkPolicyPorts
+  - Estimated increase in size: New struct inside the object with two fields of 
+  16 bits each + 16 bits for each port in the ``Exceptions`` array
```
```suggestion
  16 bits each + 16 bits for each port in the ``Except`` array
```

Dyanngg commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,434 @@
+#### Story 3 - Containerized Passive FTP Server
+As a Kubernetes User, I've received a demand from my boss to run our FTP server in an 
+existing Kubernetes cluster, to support some of my legacy applications. 
+his FTP Server must be acessible from inside the cluster and outside the cluster, 
```
This FTP Server?

Dyanngg commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,434 @@
+
+    // Except defines all the exceptions in the port range
+    +optional
+    Except       []uint16
```
Just wanted to clarify here if the expected behavior of the Except block, like in NetworkPolicy ipBlock, is fallthrough rather than explicit deny? See https://github.com/kubernetes/kubernetes/issues/83191 for more information. I'm mentioning this because if this is the case, then for the `User Story 2` you brought up in the previous section, it might be important to note this Except block will not create explicit port deny rule.

rikatz replied:
Yeah, right :)

So by the other discussions, we've actually decided to drop the Exception instead of use it. This is going to bring too much confusion for the users and can be achieved by creating different port ranges (see https://kubernetes.slack.com/archives/C01CWSHQWQJ/p1605898496104400).

The thing is, if we decide to add the exception later this might be a breaking change because CNIs not aware of this may fail open a range that you insert an exception.

By now, IMO it's better that we keep this simple and drop the Exception. Sounds good?

Dyanngg replied:
Makes sense to me

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,434 @@
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyIngressRule`` and 
```
If I read correctly this is not a compatible change.

If the user specifies `portRanges` and does not specify `ports` they hit this part of the spec:

```
type NetworkPolicyIngressRule struct {
    // List of ports which should be made accessible on the pods selected for this
    // rule. Each item in this list is combined using a logical OR. If this field is
    // empty or missing, this rule matches all ports (traffic not restricted by port).
    // If this field is present and contains at least one item, then this rule allows
    // traffic only if the traffic matches at least one port in the list.
    // +optional
    Ports []NetworkPolicyPort
```

Specifically "If this field is empty or missing, this rule matches all ports"

In other words, a back-rev implementation which does not understand the new field will fail-open, which is a no-no.

What if, instead, you added a "size" or "rangeSize" or "endPort" field to `NetworkPolicyPort` which converted the single port into a range.  That way, a back-rev implementation who did not know about the new field would allow access to/from the single "base" port"  and nothing else.  Still not the correct interpretation, but it fails closed.

I *think* that could work.  Could use extra brains on it.

rikatz replied:
Same problem from namespaceName selector, right? What if the CNI does not know the new field, got it.

I see value here in having the endPort (don't think rangeSize is very intuitive, as you need to make some 'math' to calculate big ranges)..

fasaxc replied:
Calico's datamodel supports port ranges by specifying the start and end port; this has worked well and I can't remember anyone having any issues grokking it; @thockin's idea sounds like a least-bad way to handle back compat.

Agree with @rikatz that rangeSize is not intuitive.

Could the struct be renamed to `NetworkPolicyPortRange` and still preserve back-compat (since the struct name doesn't actually feature in the JSON/YAML).

abhiraut replied:
> Calico's datamodel supports port ranges by specifying the start and end port; this has worked well and I can't remember anyone having any issues grokking it; @thockin's idea sounds like a least-bad way to handle back compat.
> 
> Agree with @rikatz that rangeSize is not intuitive.
> 
> Could the struct be renamed to `NetworkPolicyPortRange` and still preserve back-compat (since the struct name doesn't actually feature in the JSON/YAML).

renaming the struct would probably break anyone referring to the type downstream?

rikatz replied:
Yes, I guess. I'm afraid of someone referencing the struct name and breaking things. I think probably the best now is to specify the endPort (assuming the endPort is also part of the range)

fasaxc replied:
> renaming the struct would probably break anyone referring to the type downstream?

Could leave a type alias in its place.  No big deal either way.

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,434 @@
+
+### Notes/Constraints/Caveats
+
+*  The technology used by the CNI provider might not support port range in a 
```
We should cross-check this with the biggest implementations to see if it is feasible.

fasaxc replied:
FWIW, Calico has native support for port ranges (in iptables and BPF modes) so this would be straightforward for us.  iptables has native support for port ranges so any solution that uses that should be fine.

abhiraut replied:
For OpenFlow based implementations, port ranges can be achieved using masks. At least in Antrea we can adapt to this change. I believe Cilium also natively supports range of ports in CliliumNetworkPolicies.

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,434 @@
+
+    // Except defines all the exceptions in the port range
+    +optional
+    Except       []uint16
```
I'm not super convinced we need Except.  CIDRs have to be powers of 2, which makes exceptions hard.  Port ranges have no such requirement, so any exception can be easily represented as 2 ranges.

rikatz replied:
This was the original proposal, but we've discussed about the hardness of implementing this.

OTOH, by the last comment of adding the endPort field creating exceptions can be harder (unless Expections are also part of NetworkPolicyPort).

I don't have a strong opinion actually about the exception thing, so will leave for other folks to comment.

fasaxc replied:
Except clause seems like overkill to me too.  It makes validation a bit more fiddly (exceptions must lie in the range, presumably) and of course, once you have an exception ports, why not exception ranges?  From an implementer's point of view, it makes it more fiddly but not by much.

rikatz commented:
Dropped the Exception proposal, and moved to endPort as proposed by Tim.

I'm not sure we want to use the field as uint16 (as ports cannot be negative, I guess) or int32, so I kept as uint16.

aojea commented:
> I'm not sure we want to use the field as uint16 (as ports cannot be negative, I guess) or int32, so I kept as uint16.

I think that this applies https://github.com/kubernetes/community/blob/4f6942fb9d7909d004a93e274865249cf0107e96/contributors/devel/sig-architecture/api-conventions.md#primitive-types

Also, current APIs for Ports use int32

```
pkg/apis/core/types.go: HostPort int32
pkg/apis/core/types.go: ContainerPort int32
pkg/apis/core/types.go: Port intstr.IntOrString
pkg/apis/core/types.go: Port intstr.IntOrString
pkg/apis/core/types.go: HealthCheckNodePort int32
pkg/apis/core/types.go: // the 'Name' field in EndpointPort objects.
pkg/apis/core/types.go: Port int32
pkg/apis/core/types.go: TargetPort intstr.IntOrString
pkg/apis/core/types.go: NodePort int32
pkg/apis/core/types.go:// expanded set of endpoints is the Cartesian product of Addresses x Ports.
pkg/apis/core/types.go: Ports             []EndpointPort
pkg/apis/core/types.go:// EndpointPort is a tuple that describes a single port.
pkg/apis/core/types.go:type EndpointPort struct {
pkg/apis/core/types.go: Port int32
pkg/apis/core/types.go: // Port number of the given endpoint.
pkg/apis/core/types.go: Port int32
pkg/apis/core/types.go: Ports []int32
pkg/apis/core/types.go: Port int32
```

rikatz commented:
> I think that this applies
> 
> > All public integer fields MUST use the Go (u)int32 or Go (u)int64 types, not (u)int (which is ambiguous depending on target platform). Internal types may use (u)int
> 
> https://github.com/kubernetes/community/blob/4f6942fb9d7909d004a93e274865249cf0107e96/contributors/devel/sig-architecture/api-conventions.md#primitive-types

Thank you @aojea ! TIL :) Updating here

andrewsykim commented:
I think port exceptions in the API really boils down to use-cases and requirements. 

If we anticipate users will have 1 or 2 exceptions at-most, it probably makes sense to exclude it. If we think users will typically have something like 5 or more exceptions, that would require N+1 distinct ranges to express those exceptions. Users have to infer the exceptions by parsing where the each range starts and ends. This seems fine for 1-2 exceptions, but maybe not okay for more, this feels error prone if you have to deal with more than a handful of port exceptions.

rikatz commented:
> I think port exceptions in the API really boils down to use-cases and requirements.
> 
> If we anticipate users will have 1 or 2 exceptions at-most, it probably makes sense to exclude it. If we think users will typically have something like 5 or more exceptions, that would require N+1 distinct ranges to express those exceptions. Users have to infer the exceptions by parsing where the each range starts and ends. This seems fine for 1-2 exceptions, but maybe not okay for more, this feels error prone if you have to deal with more than a handful of port exceptions.

So in my case, the default blocking policy covers ~8 ports (file sharing, r* commands, xterm) so I think having 8 port ranges is 'plausible' (using the user story of "I can allow all traffic from my Pods but the specific ports that exists in the company security policy").

rikatz commented:
@thockin Are you/can you be assigned as the approver of this one? If so I'll change in the kep.yaml and wait for more considerations.

thockin commented:
Yeah, you can make me approver

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,383 @@
+
+## Summary
+
+Today the ``ports``field in ingress and egress network policies is an array 
```
You only need a single backtick

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,383 @@
+	// +optional
+	Port       *intstr.IntOrString `json:"port,omitempty" protobuf:"bytes,2,opt,name=port"`
+
+	// EndPort defines the end of the port range, being the end included within the
```
In terms of wording maybe:

`EndPort defines the last port included in the port range.`

and maybe an example?

rikatz replied:
Ok, I've inserted  a 'simple' example. When you say example is in the code (that turns into docs) right? Don't know if this is right, will take a look into other parts of k/k code late tomorrow

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,383 @@
+### Validations
+The ``NetworkPolicyPort`` will need to be validated, with the following scenarios:
+* If an ``EndPort`` is specified a ``Port`` must also be specified
+* If ``Port`` is a string ``EndPort`` cannot be specified
```
I think this limitation is OK, but not strictly required

abhiraut replied:
what would be the interpretation of 

```
Port: http
EndPort: 9000
```
?
maybe it is better to take a stand and not allow this?

rikatz replied:
+1 not to allowing this. This can lead to a conversion confusion to CNIs. One concern I had was: does iptables supports named ports? Otherwise we could assume those ports are always converted to int by the CNI before inserting into the dataplane.

The answer is: yes, it accepts. Because we don't know how this is done in the wild (if there's some CNI creating rules with the named ports, and not the numbered ones) I think it's better also to say 'if you want a range, you can only use integer as the start of the range'

thockin replied:
Whatever implements policy must convert named ports to numeric.

We *could* allow range here but it feels like a super corner case and let's just not.

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,383 @@
+
+### Validations
+The ``NetworkPolicyPort`` will need to be validated, with the following scenarios:
+* If an ``EndPort`` is specified a ``Port`` must also be specified
```
`EndPort` must be >= `Port` (note that == is allowed - is that OK?)

rikatz replied:
great, forgot this case. IMO it's ok, == represents a single port. Don't know why someone would like to have a Port and an EndPort equal (compliance? etc) but seems reasonable to me

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,383 @@
+
+If downgraded the CNI wont be able to look into the new field, as this does not 
+exists and network policies using this field will stop working correctly and
+start working incorrectly.
```
but they will fail-closed so it is acceptable

thockin commented on keps/sig-network/2079-network-policy-port-range/README.md:
```diff
@@ -0,0 +1,386 @@
+
+	// EndPort defines the last port included in the port range.
+	// Example:
+  //    endPort: 12345
```
for example I meant something like `{ port: 8000, endport: 8002 } // ports 8000, 8001, 8002`

thockin replied:
maybe an example is dumb here, don't sweat it. :)

thockin reviewed (commented):
Content-wise I am OK with this.  Let's iterate.

Thanks!

/lgtm
/approve
